packages/devtools_app/lib/src/screens/performance/timeline_event_processor.dart

```diff
@@ -366,40 +342,98 @@ class TimelineEventProcessor {
+          //
+          debugTraceEventCallback(
+            () => log(
+              'Duplicate duration begin event - removing duplicate:'
+              ' ${current?.beginTraceEventJson}',
+            ),
+          );
+
+          current.parent!.removeChild(current);
+          current = current.parent as SyncTimelineEvent?;
+          currentDurationEventNodes[eventThreadId] = current;
+        } else {
+          // We have a valid begin event that is missing a matching end event.
+          // Create a fake end event to re-balance the tree.
+          //
+          // Trace example:
+          // VSYNC - DurationBegin
+          // BUILD - DurationBegin [current] (missing a matching end event)
+          // Animator::BeginFrame - DurationBegin
+          // ...
+          // Animator::BeginFrame - DurationEnd [previousDurationEndEvent]
+          // <this is where the DurationEnd BUILD event should be>
+          // VSYNC - DurationEnd [event]
+          //
+          final currentBeginTrace = current.traceEvents.first.event;
+
+          // For the fake end timestamp, split the difference between the end
+          // time of the last child of [current] and the end time of the [event]
+          // we are currently trying to process. Or, if [current] has no
+          // children, use the end timestamp [event] with a small buffer to
+          // ensure the fake event ends before the [event].
+          final lastChildOfCurrentEndTime = current
+              .children.safeLast?.endTraceEventJson?[TraceEvent.timestampKey];
+          final eventEndTime = event.timestampMicros!;
+          const fakeEndEventTimeBuffer = 1;
+          final fakeTimestampMicros = lastChildOfCurrentEndTime != null
+              ? eventEndTime -
+                  ((eventEndTime - lastChildOfCurrentEndTime) / 2).round()
+              : eventEndTime - fakeEndEventTimeBuffer;
+
+          final fakeEndEvent = TraceEventWrapper(
+            currentBeginTrace.copy(
+              phase: TraceEvent.durationEndPhase,
+              timestampMicros: fakeTimestampMicros,
+              args: {
+                'message': 'Warning - the end time of this event may be '
+                    'innacurate. The end trace event was missing, so the end '
+                    'time was inferred.',
+              },
+            ),
+            // Use the same time received as the current event we are trying to
+            // process.
+            eventWrapper.timeReceived,
           );
-        }
 
-        current.parent!.removeChild(current);
-        current = current.parent as SyncTimelineEvent?;
-        currentDurationEventNodes[eventThreadId] = current;
+          debugTraceEventCallback(() {
+            log(
+              'Missing DurationEnd event - adding fake end event $fakeEndEvent',
+            );
+          });
+
+          current.addEndEvent(fakeEndEvent);
+          current = current.parent as SyncTimelineEvent?;
```

Comment:
Is this cast necessary?

Reply:
yeah. without it, the following analysis error is present: "A value of type 'TimelineEvent?' can't be assigned to a variable of type 'SyncTimelineEvent?'"